### ISSUE TITLE
How to get notified on unsubscribe / drop connection

### ISSUE THREAD

**Issue by xlc** [paritytech/jsonrpsee#990]
https://github.com/paritytech/jsonrpsee/blob/dab1bfc0eee59a06feedeb3c2593c57157b33ca2/core/src/server/rpc_module.rs#L828

I can't find any public API to expose this to allow me cleanup resources on drop connection. Also no public API to notify unsubscribe as well.

**Comment by niklasad1** [comment]
There is only https://github.com/paritytech/jsonrpsee/blob/dab1bfc0eee59a06feedeb3c2593c57157b33ca2/core/src/server/rpc_module.rs#L104, I guess we could have an async fn `closed` I have added that in a pending PR to add backpressure.

Thus, it's on it's way

**Comment by lexnv** [comment]
An alternative to that, depending on your use-case, would be to interact with the subscription.

Let's presume the client has disconnected or called the unsubscribed method.
Presume the server doesn't use the `is_closed` or any other future to detect the disconnect, then the server will eventually try to submit a message to the subscription. This is where the server receives `Ok(false)` as a result of that action. 

https://github.com/paritytech/jsonrpsee/blob/dab1bfc0eee59a06feedeb3c2593c57157b33ca2/core/src/server/rpc_module.rs#L893-L900


To that regard, there is also this suggestion in discussion: https://github.com/paritytech/jsonrpsee/issues/944, which would make things a bit more explicit to the user. Currently, you can shoot yourself in the foot with a code of those lines:
```rust
loop {
  match s.send() {
    Ok(_) => // if it is false, then we should stop.
  ..
}}
```

**Comment by xlc** [comment]
My use case is a JSON RPC proxy that forwards subscriptions to upstream server. So I will want to unsubscribe from upstream as soon as the downstream disconnect/unsubscribe.

So I don't want to use the sync closed method because when should I check it? and dont' want to only drop subscription on failed to send because that will just be a waste of resources on both proxy server and upstream server.

**Comment by niklasad1** [comment]
Yeah, that makes sense we should provide an async function that returns once this connection is closed or subscription is canceled.

It's useful for folks to build their own async primitives instead of using `pipe_from_stream`

My PR #962 will enable that.

**Comment by niklasad1** [comment]
Closed by #962 

We have now introduced `SubscriptionSink::closed` that notifies once the subscription is closed.
